scene/resources/visual_shader_particle_nodes.cpp

```diff
@@ -1162,6 +1162,205 @@ VisualShaderNodeParticleAccelerator::VisualShaderNodeParticleAccelerator() {
 	simple_decl = false;
 }
 
+// VisualShaderNodeParticleTurbulence
+
+String VisualShaderNodeParticleTurbulence::get_caption() const {
+	return "ParticleTurbulence";
+}
+
+int VisualShaderNodeParticleTurbulence::get_output_port_count() const {
+	return 2;
+}
+
+VisualShaderNodeParticleTurbulence::PortType VisualShaderNodeParticleTurbulence::get_output_port_type(int p_port) const {
+	switch (p_port) {
+		case 0:
+			return PORT_TYPE_SCALAR;
+		case 1:
+			return PORT_TYPE_VECTOR_3D;
+	}
+	return PORT_TYPE_SCALAR;
+}
+
+String VisualShaderNodeParticleTurbulence::get_output_port_name(int p_port) const {
+	switch (p_port) {
+		case 0:
+			return "turbulence influence";
+		case 1:
+			return "noise direction";
+	}
+	return String();
+}
+
+bool VisualShaderNodeParticleTurbulence::is_input_port_default(int p_port, Shader::Mode p_mode) const {
+	return p_port == 1; // seed
+}
+
+int VisualShaderNodeParticleTurbulence::get_input_port_count() const {
+	return 9;
+}
+
+VisualShaderNodeParticleTurbulence::PortType VisualShaderNodeParticleTurbulence::get_input_port_type(int p_port) const {
+	switch (p_port) {
+		case 0:
+			return PORT_TYPE_VECTOR_3D;
+		case 1:
+			return PORT_TYPE_SCALAR_UINT;
+		case 2:
+			return PORT_TYPE_SCALAR;
+		case 3:
+			return PORT_TYPE_SCALAR;
+		case 4:
+			return PORT_TYPE_VECTOR_3D;
+		case 5:
+			return PORT_TYPE_SCALAR;
+		case 6:
+			return PORT_TYPE_SCALAR;
+		case 7:
+			return PORT_TYPE_SCALAR;
+		case 8:
+			return PORT_TYPE_SCALAR;
+	}
+	return PORT_TYPE_SCALAR;
+}
+
+String VisualShaderNodeParticleTurbulence::get_input_port_name(int p_port) const {
+	switch (p_port) {
+		case 0:
+			return "particle position";
+		case 1:
+			return "noise seed";
+		case 2:
+			return "noise strength";
+		case 3:
+			return "noise scale";
+		case 4:
+			return "noise speed";
+		case 5:
+			return "noise speed random";
+		case 6:
+			return "influence min";
+		case 7:
+			return "influence max";
+		case 8:
+			return "influence over life";
+	}
+	return String();
+}
+
+String VisualShaderNodeParticleTurbulence::generate_global_per_node(Shader::Mode p_mode, int p_id) const {
+	String code;
+
+	code += "vec4 turb_grad(vec4 p) {\n";
```

Comment:
I'd suggest doing this with R strings

Reply:
Could you give me an example or point me to a file where it is implemented with R strings? Haven't used R strings yet

Reply:
See `sky_material` for an example 🙂

Reply:
Wait, you mean for `generate_global_per_node`? Why there? If anywhere I would assume `generate_code`

Reply:
Because it's a long block of strings being added and it's much cleaner if it's in one block of string instead

Reply:
Should be done now :D

Reply:
Allowed myself to mark this as resolved :)